rfc/expr-context.md

```diff
@@ -0,0 +1,134 @@
+# Expression Context
+
+Cosmo Router provides a context object to an expr-lang VM that can be used to access information about the request, response, and other relevant information. The context object is available at all times during the request's lifecycle.
+
+## request
+
+The client request object is a read-only object that provides information about the incoming request. It is available at all times during the request's lifecycle.
+
+```
+request.id
+request.header
+request.method
+request.body.query
+request.body.operation_name
+request.body.variables
+request.body.extensions
+request.uri.host
+request.uri.path
+request.uri.port
+
+request.client.name
+request.client.version
+request.client.ip
+request.routerConfigVersion
+```
+
+# request.context
+
+The context object provides information about the context of the request. A generic object that can be used to store any information that is relevant to the request.
+
+```
+request.context.agent
+```
+
+In the Router config, you can use the following expression to set a custom field in the context object:
+
+```yaml
+context:
+  agent: "request.auth.claims.sub || request.client.ip || request.header.X-Forwarded-For"
+```
+
+At runtime, the following expression can be used to set additional fields in the context object or to override or unset existing fields:
+
+```
+# set agent to undefined
+unset("request.context.agent")
+```
+
+```
+# set agent to request.auth.claims.sub
+set("request.context.agent", request.auth.claims.sub)
+```
+
+```
+# add another field to the context object
+set("request.context.client", request.client.name)
+```
+
+# authentication
+
+The authentication object provides information about the authentication of the request.
+
+```
+request.isAuthenticated // true or false
+request.auth // Only set when authenticated
+request.auth.type
+request.auth.claims
+request.auth.scopes
+```
+
+# request.operation
+
+The operation object provides information about the operation being executed. This information is parsed and validated by the router.
+
+```
+request.operation.name // e.g. MyQuery
+request.operation.type // query, mutation, subscription
+request.operation.hash // hash of the operation
+```
+
+# fetch.*
```

Comment:
I'm not a fan of the fetch terminology. It's not used anywhere on the server world. It always reminds me to the browser.

Reply:
Everything from Router to Subgraphs in the Federation space is called fetch

Reply:
I understand but because federation is protocol agnostic. Here we clearly reference to http primitives like headers.